Ash.FrontEnd/src/js/global/components/Fido.jsx

```diff
@@ -0,0 +1,293 @@
+        },
+        {}
+      );
+      this.setState({ ...createInitFetchState(newFetchItems, this.onCall) });
+    }
+  }
+
+  /**
+   * Get the configurations derived from fetchItems
+   * Small wrapper for *generateConfigs*
+   */
+  getConfigs = () => {
+    const { fetchItems } = this.props;
+
+    return this.generateConfigs(fetchItems);
+  };
+
+  /**
+   * Generate configurations for the supplied fetches
+   * **Memoized** so that configs are only regenerated if fetchItems changes
+   * @param {object} fetchItems - Object containing named config objects
+   * @returns {object} Named config objects composed of the default config
+   *    and the supplied config object
+   */
+  generateConfigs = memoizeOne(fetchItems =>
+    Object.entries(fetchItems).reduce((configs, [key, config]) => {
+      configs[key] = { ...defaultFetchConfig, ...config };
+
+      return configs;
+    }, {})
+  );
+
+  /**
+   * Call method for initiating a fetch
+   * @param {string} key - fetch item key
+   * @returns {(configOverride: object) => void} accepts immediate overrides to the
+   *    config object and calls the dispatchFetch method
+   */
+  onCall = key => configOverride => {
+    const configs = this.getConfigs();
+
+    let newConfig = configs[key];
+    if (configOverride) {
+      newConfig = { ...config, ...configOverride };
+    }
+
+    this.dispatchFetch(key, newConfig);
+  };
+
+  /**
+   * Initiate a fetch with a given configuration
+   * @param {string} key - fetch item key
+   * @param {object} config - fetch item configuration
+   */
+  dispatchFetch = (key, config) => {
+    const method = config.options.method.toUpperCase();
+    const abortController = new AbortController();
+
+    let body = null;
+    if (method !== 'GET' && method !== 'HEAD') {
+      body = JSON.stringify(config.payload);
+    }
+
+    let headers = new Headers();
+    if (body) headers.append('Content-Type', 'application/json');
+
+    // Abort previous fetch if there is any
+    if (this.aborts[key]) this.aborts[key].abort();
+    // Replace previously aborted (or new) AbortController
+    this.aborts[key] = abortController;
+
+    let fetchUrl = config.path;
+    if (config.query) {
+      const stringifiedQueryParams = Object.entries(config.query)
+        .map(([key, param]) => `${key}=${param}`)
+        .join('&');
+      fetchUrl += `?${stringifiedQueryParams}`;
+    }
+
+    fetch(fetchUrl, {
+      method,
+      headers,
+      body,
+      mode: 'same-origin', // only allow traffic to our back end
+      credentials: 'same-origin',
+      cache: 'no-store', // do not cache fetches made from this component
+      redirect: 'follow', // default
+      referrer: 'client', // default
+      referrerPolicy: 'no-referrer-when-downgrade', // default
+      // integrity
+      // keepalive
+      signal: abortController.signal,
+    })
+      .then(Fido.processResponse)
+      .then(this.setData(key))
+      .catch(this.onFailure(key));
```

Comment:
Be very wary of this.
Currently the way you have things set, if either Fido.processResponse or this.setData() go trigger client code(IE someone using Fido), this catch statement will catch errors in their code as well.

This causes some extremely weird bugs, where bad renders/etc from client are being caught here and act like it's Fido's problem.

It might be you have this handled, it was just a big issue for my component I made lol, had to edit it

Reply:
Looking at what you have it looks like it might be okay, just be aware of that issue

Reply:
Hmm, I think I understand what you are saying.  My motivation, obviously, was to catch errors on the fetch promise only.  I guess we'll have to keep an eye on it.